pkg/cloudclient/aws/private.go

```diff
@@ -220,14 +221,30 @@ func (c Client) findUnreachableEndpoints(ctx context.Context, instanceID string)
 	err := wait.PollImmediate(30*time.Second, 10*time.Minute, func() (bool, error) {
 		output, err := c.ec2Client.GetConsoleOutput(ctx, &ec2.GetConsoleOutputInput{InstanceId: &instanceID})
 		if err == nil && output.Output != nil {
-			// Find unreachable targets from output
+			// First, gather the ec2 console output
 			scriptOutput, err := base64.StdEncoding.DecodeString(*output.Output)
 			if err != nil {
 				// unable to decode output. we will try again
+				c.logger.Debug(ctx, "Error while collecting console output, will retry on next check interval: %s", err)
 				return false, nil
 			}
 
-			// If debug logging is enabled, output the full console log
+			// In the early stages, an ec2 instance may be running but the console is not populated with any data, retry if that is the case
+			if len(scriptOutput) < 1 {
+				c.logger.Debug(ctx, "EC2 console output not yet populated with data, continuing to wait...")
+				return false, nil
+			}
+
+			// Check for the specific string we output in the gerated userdata file at the end to verify the userdata script has run
+			// It is possible we get EC2 console output, but the userdata script has not yet completed.
+			reVerify := regexp.MustCompile(userdataEndVerifier)
+			verifyMatch := reVerify.FindString(string(scriptOutput))
+			if len(verifyMatch) < 1 {
```

Comment:
As things stand, if we hit this we should probably just error out, as it seems the log won't ever be updated if the first update does not include the userdata output.